### ISSUE TITLE
Router is not loading correctly

### ISSUE THREAD

**Issue by jamesalways** [NickMcBurney/storybook-vue3-router#54]
Hi! I just updated everything to the new version of Storybook and this package and something weird is happening:
- If I refresh the page the `vueRouter()` decorator is not loaded and router components and `useRouter` composable etc. don't exist.
- If I navigate to another story everything works fine, then If I refresh the page I have the same errors again

Errors:
- `[Vue warn]: injection "Symbol(router)" not found. `
- After `useRoute` when I do for instance `route.beforeEach`: `runtime.js:4 TypeError: Cannot read properties of undefined (reading 'beforeEach')`
- `[Vue warn]: Failed to resolve component: router-link` and more error related

```
"storybook": "^7.3.2",
"storybook-vue3-router": "^3.0.0",
```

Example of a story:
```
import { vueRouter } from 'storybook-vue3-router'
import { action } from '@storybook/addon-actions'
import { LINKS, CUSTOM_ROUTES } from '../mockdata'
import MainMenu from './MainMenu.vue'

export default {
  title: 'Navigation/MainMenu',
  component: MainMenu
}

export const Default = () => ({
  components: { MainMenu },
  data: () => ({
    ...LINKS
  }),
  methods: {
    handleAction (event) {
      action('Event:')(event)
    }
  },
  template: `
    <div class="bg-white min-h-screen flex">
      <MainMenu
        :logoLink="logoLink"
        :links="links"
        @click="handleAction('click')"
      />
      <div class="flex-1">
        <RouterView />
      </div>
    </div>
  `
})

Default.decorators = [
  vueRouter(CUSTOM_ROUTES)
]

```

UPDATE: 

With Storybook version `7.0.25` works fine. After `^7.1.0` stop working well..

**Comment by tfoxy** [comment]
Hi @jamesalways !

I had the same problem that you are mentioning. I discovered that this happens because `setup` callback will not be called when story is already being rendered. Now that I understand Storybook better I created a decorator:

```ts
import { Decorator, setup } from '@storybook/vue3'
import { App } from 'vue'
import { RouterOptions, createRouter, createWebHashHistory } from 'vue-router'

const Home = { template: '<div>Home</div>' }
const About = { template: '<div>About</div>' }

export const routerOptions: RouterOptions = {
  history: createWebHashHistory(),
  routes: [
    { path: '/', component: Home },
    { path: '/about', component: About },
  ],
}

export default function withRouter(options?: RouterOptions): Decorator {
  let app: App

  setup((setupApp) => {
    app = setupApp
  })

  return (story) => {
    const router = createRouter({ ...routerOptions, ...options })
    app.use(router)
    return story()
  }
}
```

You can create a PR where you can call `setup` before `makeDecorator`.

https://github.com/NickMcBurney/storybook-vue3-router/blob/b8ecc62347368b0066ad88b28081e5b158086be7/src/withVueRouter.ts#L42

```ts
) =>
  let app: App;
  setup((setupApp) => {
    app = setupApp
  })
  makeDecorator({
    name: 'withVueRouter',
    parameterName: 'withVueRouter',

    wrapper: (storyFn: StoryFn, context: StoryContext) => {
      /* setup router var */
      let router
```

**Comment by NickMcBurney** [comment]
> Hi @jamesalways !
> 
> I had the same problem that you are mentioning. I discovered that this happens because `setup` callback will not be called when story is already being rendered. Now that I understand Storybook better I created a decorator:
> 
> ```ts
> import { Decorator, setup } from '@storybook/vue3'
> import { App } from 'vue'
> import { RouterOptions, createRouter, createWebHashHistory } from 'vue-router'
> 
> const Home = { template: '<div>Home</div>' }
> const About = { template: '<div>About</div>' }
> 
> export const routerOptions: RouterOptions = {
>   history: createWebHashHistory(),
>   routes: [
>     { path: '/', component: Home },
>     { path: '/about', component: About },
>   ],
> }
> 
> export default function withRouter(options?: RouterOptions): Decorator {
>   let app: App
> 
>   setup((setupApp) => {
>     app = setupApp
>   })
> 
>   return (story) => {
>     const router = createRouter({ ...routerOptions, ...options })
>     app.use(router)
>     return story()
>   }
> }
> ```
> 
> You can create a PR where you can call `setup` before `makeDecorator`.
> 
> https://github.com/NickMcBurney/storybook-vue3-router/blob/b8ecc62347368b0066ad88b28081e5b158086be7/src/withVueRouter.ts#L42
> 
> ```ts
> ) =>
>   let app: App;
>   setup((setupApp) => {
>     app = setupApp
>   })
>   makeDecorator({
>     name: 'withVueRouter',
>     parameterName: 'withVueRouter',
> 
>     wrapper: (storyFn: StoryFn, context: StoryContext) => {
>       /* setup router var */
>       let router
> ```

Thanks @tfoxy, I'll give this a go

**Comment by tfoxy** [comment]
Hi @NickMcBurney ! I realized that the previous decorator isn't working properly when changing values in the controls. I fixed it by calling `app.use` inside decorator `setup`:

```ts
import { Decorator, setup } from '@storybook/vue3'
import { App } from 'vue'
import { RouterOptions, createRouter, createWebHashHistory } from 'vue-router'

const Home = { template: '<div>Home</div>' }
const About = { template: '<div>About</div>' }

export const routerOptions: RouterOptions = {
  history: createWebHashHistory(),
  routes: [
    { path: '/', name: 'home', component: Home },
    { path: '/about', name: 'about', component: About },
  ],
}

export default function withRouter(options?: RouterOptions): Decorator {
  let app: App

  setup((setupApp) => {
    app = setupApp
  })

  return () => {
    return {
      setup() {
        const router = createRouter({ ...routerOptions, ...options })
        app.use(router)
      },
      template: '<story/>',
    }
  }
}
```

**Comment by jamesalways** [comment]
The fix works perfectly. Thanks!!

I updated the file to match the library interface and use the library when the fix is ready again.

````
import { Decorator, setup } from '@storybook/vue3'
import { App } from 'vue'
import { RouteRecordRaw, RouterOptions, createRouter, createWebHashHistory } from 'vue-router'

const Home = { template: '<div>Home</div>' }
const About = { template: '<div>About</div>' }

export const routerOptions: RouterOptions = {
  history: createWebHashHistory(),
  routes: [
    { path: '/', name: 'home', component: Home },
    { path: '/about', name: 'about', component: About }
  ]
}

export function vueRouter (routes?: RouteRecordRaw[]): Decorator {
  let app: App

  setup((setupApp) => {
    app = setupApp
  })

  return () => ({
    setup () {
      const router = createRouter({ ...routerOptions, routes: routes || routerOptions.routes })
      app.use(router)
    },
    template: '<story/>'
  })
}

**Comment by JoepRoggekamp** [comment]
@jamesalways We are running into the same issue and I'd like to resolve this issue for my team. Where/How can we apply the fix you mentioned above?

**Comment by jamesalways** [comment]
> @jamesalways We are running into the same issue and I'd like to resolve this issue for my team. Where/How can we apply the fix you mentioned above?

@JoepRoggekamp 

- Create a file with the previous code wherever you want.
- In the story, import and use it like this:

````js
import { vueRouter } from 'yourPathToTheFile'

StoryName.decorators = [
  vueRouter() or   vueRouter([...routes])
]

**Comment by NickMcBurney** [comment]
FYI I closed then I merged fix, but auto publish script didnt work so I still need to resolve that. Should have this sorted this evening though and ill confirm version here.

**Comment by NickMcBurney** [comment]
Issue resolved and packaged publish: 4.0.1

**Comment by dlferro** [comment]
Out of curiosity, has anyone run into this same issue while using autodocs? Individually each story pages appear to render without issue.

**Comment by zenatshockbyte** [comment]
> Out of curiosity, has anyone run into this same issue while using autodocs when you have multiple stories? Individually each story pages appear to render without issue.

Yeah this issue is happening for us on the docs page, not for individual stories

**Comment by zejji** [comment]
@dlferro  - Yes, I have run into this issue on the docs page for a component when it has multiple stories. One of the stories renders correctly, but all the others trigger errors as describe above.

I think this issue should be reopened.

Unfortunately I'm a newcomer to Storybook and haven't yet been able to identify a fix.

**Comment by pa-bouly** [comment]
Hey @zejji 
in this case, i think it's because the vue-router is set up multiple times.
So if you use the code shown previously, you can add a condition to use the vue-router like so : 

```javascript
  return () => ({
    setup() {
      if (app._context.components['RouterLink'] === undefined) {
        const router = createRouter({ ...routerOptions, routes: routes || routerOptions.routes })
        app.use(router)
      }
    },
    template: '<story/>',
  })
```

**Comment by tfoxy** [comment]
Hey @dlferro @zejji @pa-bouly I created #56 to fix this issue. Let me know if it works for you so that it can be merged without causing any other issue.

**Comment by NickMcBurney** [comment]
Thanks @tfoxy, I've tested, merged PR and published new version. 

My Release Pipeline all messed up so managed to publish version 4.1.0 (initial version with fix) all the way to 4.25.0, and I can unpublish duplicate versions due to some package have this listed as a dependancy - with some very loose version requirement which means all versions are considered a dependency and blocked from unpublishing.
